File: December2022/Java/OddEvenLinkedList.java
// Problem Link: https://leetcode.com/problems/odd-even-linked-list/description/

// author: Mahesh Reddy B N

class Solution {
    public ListNode oddEvenList(ListNode head) {
    if(head ==null || head.next ==null || head.next.next == null) return head;
    ListNode odd = head;
    ListNode even = head.next;
    ListNode evenStart = head.next;
    while( odd.next!= null && even.next != null  ){
        odd.next = even.next;
        odd = odd.next;
        even.next = odd.next;
        even = even.next;
    }
    odd.next = evenStart;
    return head;
    }
}

File: December2022/C++/OddEvenLinkedList.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* oddEvenList(ListNode* head) {
        if(head==nullptr) return head;
        ListNode *d1=head,*d2=head->next,*d=head->next;
        while(d2 && d2->next)
        {
            d1->next = d1->next->next;
            d1 = d1->next;
            d2->next = d2->next->next;
            d2 = d2->next;
        }
        d1->next = d;
        return head;
    }
};